reconcile/test/test_terraform_tgw_attachments.py

```diff
@@ -0,0 +1,302 @@
+        "spec": {
+            "region": "us-east-1",
+        },
+        "network": {"vpc": "10.0.0.1/16"},
+        "peering": {
+            "connections": [
+                account_vpc_connection,
+            ]
+        },
+    }
+
+
+@pytest.fixture
+def cluster_with_mixed_connections(account_tgw_connection, account_vpc_connection):
+    return {
+        "name": "cluster_with_mixed_connections",
+        "ocm": {"name": "cluster_with_mixed_connections-ocm"},
+        "spec": {
+            "region": "us-east-1",
+        },
+        "network": {"vpc": "10.0.0.2/16"},
+        "peering": {
+            "connections": [
+                account_tgw_connection,
+                account_vpc_connection,
+            ]
+        },
+    }
+
+
+@pytest.fixture
+def tgw():
+    return {
+        "tgw_id": "tgw-1",
+        "tgw_arn": "tgw-arn-1",
+        "region": "us-west-1",
+        "routes": [],
+        "rules": [],
+        "hostedzones": None,
+    }
+
+
+@pytest.fixture
+def vpc_details():
+    return {
+        "vpc_id": "vpc-id-1",
+        "route_table_ids": ["rtb-1"],
+        "subnets_id_az": [{"az": "us-east-1a", "id": "subnet-1"}],
+    }
+
+
+@pytest.fixture
+def assume_role():
+    return "some-role"
+
+
+def _setup_mocks(
+    mocker, clusters=None, accounts=None, vpc=None, tgws=None, assume_role=None
+):
+    mocker.patch("reconcile.queries.get_secret_reader_settings", return_value={})
+    mocker.patch(
+        "reconcile.queries.get_clusters_with_peering_settings",
+        return_value=clusters or [],
+    )
+    mocker.patch("reconcile.queries.get_aws_accounts", return_value=accounts or [])
+    mocked_aws_api = mocker.patch(
+        "reconcile.terraform_tgw_attachments.AWSApi", autospec=True
+    ).return_value
+    with mocked_aws_api as aws_api:
+        aws_api.get_cluster_vpc_details.return_value = vpc or (None, None, None)
+        aws_api.get_tgws_details.return_value = tgws or []
+    mocked_ocm = mocker.patch(
+        "reconcile.terraform_tgw_attachments.OCMMap", autospec=True
+    ).return_value
+    mocked_ocm.get.return_value.get_aws_infrastructure_access_terraform_assume_role.return_value = (
+        assume_role
+    )
+    mocked_ts = mocker.patch(
+        "reconcile.terraform_tgw_attachments.Terrascript", autospec=True
+    ).return_value
+    mocked_ts.dump.return_value = []
+
+    mocked_tf = mocker.patch(
+        "reconcile.terraform_tgw_attachments.Terraform", autospec=True
+    ).return_value
+    mocked_tf.plan.return_value = (False, False)
+    mocked_tf.apply.return_value = False
+    return {
+        "tf": mocked_tf,
+        "ts": mocked_ts,
+    }
+
+
+def test_dry_run(mocker):
+    mocks = _setup_mocks(mocker)
+
+    integ.run(True, enable_deletion=False)
+
+    mocks["tf"].plan.assert_called_once_with(False)
+    mocks["tf"].apply.assert_not_called()
+
+
+def test_non_dry_run(mocker):
+    mocks = _setup_mocks(mocker)
+
+    integ.run(False, enable_deletion=False)
```

Comment:
Im wondering if instead of testing the whole `run()` method it might make sense to test single functions to reduce the whole `patch()` madness. `build_desired_state_tgw_attachments()` might be an ideal candidate for testing as it covers a lot of logic and we could pass the deps in there as clean mocks. In a next step for further tests, it might make sense to refactor the `run()` method and extract essential logic into functions with dep inversion in mind. (Personally, I really dislike python's patch approach and feel it is a sign of bad code structure in the integration itself. I know we do this in a lot of places, but if there is a way to avoid it then Im always voting for that :sweat_smile: ) Wdyt?

Reply:
I'm testing `run()` because that's the public method used eventually. I'm going to add sharding logic to it, so want to skip testing inner methods for now as they may change a lot during the process. Unless those are also public methods and we see them as interfaces. I'd like to treat them as internal sociable and external solitary. [Solitary or Sociable?](https://martinfowler.com/bliki/UnitTest.html)

I also don't like mocker patch this way, it's coupling with how I import a lib, if I change the import alias, it will break, or even work in a surprise way.

Change the existing structure to DI may not be easy, as some dependencies require results from previous step to build. If we want to have some clean tests, then we can extract a layer of pure functions, they can use clean mocks. Testing control logic still require current mocker patch as we are not using any DI framework.